test/distributed/test_utils.py

```diff
@@ -14,23 +14,210 @@
 
 import os
 
-from modulus.distributed import DistributedManager, gather_loss
+import pytest
+import torch
+import torch.nn as nn
 
+from modulus.distributed import (
+    DistributedManager,
+    gather_loss,
+    mark_module_as_shared,
+    unmark_module_as_shared,
+)
+from modulus.distributed.utils import _reduce
 
-# TODO: Need to figure out how to test parallel set up
-def test_gather_loss():
+
+def run_test_gather_loss(rank, world_size):
+    os.environ["RANK"] = f"{rank}"
+    os.environ["WORLD_SIZE"] = f"{world_size}"
     os.environ["MASTER_ADDR"] = "localhost"
-    os.environ["MASTER_PORT"] = "12345"
-    os.environ["RANK"] = "0"
-    os.environ["WORLD_SIZE"] = "1"
+    os.environ["MASTER_PORT"] = str(12355)
+
     # Reset class state
     DistributedManager._shared_state = {}
     DistributedManager.initialize()
 
     manager = DistributedManager()
     assert manager.is_initialized()
 
-    loss = gather_loss(1.0, dst_rank=0, mean=True)
-    assert loss == 1.0
+    loss = gather_loss(1.0, dst_rank=0, mean=False)
+    if manager.local_rank == 0:
+        assert loss == 1.0 * world_size, str(loss)
+    else:
+        assert True
+
     del os.environ["RANK"]
     del os.environ["WORLD_SIZE"]
+    del os.environ["MASTER_ADDR"]
+    del os.environ["MASTER_PORT"]
+    DistributedManager.cleanup()
+
+
+def run_test_mark_shared(rank, world_size):
+    class TestModule(nn.Module):
+        def __init__(self):
+            super().__init__()
+            self.lin_1 = nn.Linear(4, 2)
+            self.lin_2 = nn.Linear(2, 4)
+
+        def forward(self, x):
+            return torch.sigmoid(self.lin_2(torch.tanh(self.lin_1(x))))
+
+    os.environ["RANK"] = f"{rank}"
+    os.environ["WORLD_SIZE"] = f"{world_size}"
+    os.environ["MASTER_ADDR"] = "localhost"
+    os.environ["MASTER_PORT"] = str(12355)
+
+    DistributedManager._shared_state = {}
+    DistributedManager.initialize()
+    DistributedManager.create_process_subgroup(
+        name="shared_parallel",
+        size=world_size,
+    )
+    manager = DistributedManager()
+    assert manager.is_initialized()
+
+    torch.manual_seed(42 * world_size + rank)
+    ref_module = TestModule().to(device=manager.device)
+    torch.manual_seed(42 * world_size + rank)
+    dist_module = TestModule().to(device=manager.device)
+    x = torch.ones(4, device=manager.device)
+    ref_out = ref_module(x)
+    ref_out.backward(torch.ones_like(ref_out))
+    ref_lin_1_weight_grad = _reduce(
+        ref_module.lin_1.weight.grad.clone().detach(),
+        group=manager.group("shared_parallel"),
+        use_fp32=True,
+    )
+    ref_lin_1_bias_grad = _reduce(
+        ref_module.lin_1.bias.grad.clone().detach(),
+        group=manager.group("shared_parallel"),
+        use_fp32=True,
+    )
+    ref_lin_2_weight_grad = _reduce(
```

Comment:
Nit: Can lin_2 reductions be moved to below the first set of asserts to make the test a bit easier to follow?